src/lib.rs

```diff
@@ -1330,116 +1334,93 @@ impl Build {
             let mut stdout = io::BufWriter::with_capacity(128, io::stdout());
-            let mut backoff_cnt = 0;
 
             loop {
-                let mut has_made_progress = false;
                 // If the other end of the pipe is already disconnected, then we're not gonna get any new jobs,
                 // so it doesn't make sense to reuse the tokens; in fact,
                 // releasing them as soon as possible (once we know that the other end is disconnected) is beneficial.
                 // Imagine that the last file built takes an hour to finish; in this scenario,
                 // by not releasing the tokens before that last file is done we would effectively block other processes from
                 // starting sooner - even though we only need one token for that last file, not N others that were acquired.
-                let mut is_disconnected = false;
-                // Reading new pending tasks
-                loop {
-                    match rx.try_recv() {
-                        Ok(pending) => {
-                            has_made_progress = true;
-                            pendings.push(pending)
-                        }
-                        Err(mpsc::TryRecvError::Disconnected) if pendings.is_empty() => {
-                            let _ = stdout.flush();
-                            return if let Some(err) = error {
-                                Err(err)
-                            } else {
-                                Ok(())
-                            };
-                        }
-                        Err(mpsc::TryRecvError::Disconnected) => {
-                            is_disconnected = true;
-                            break;
-                        }
-                        _ => break,
-                    }
-                }
 
-                // Try waiting on them.
-                retain_unordered_mut(&mut pendings, |(cmd, program, child, token)| {
-                    match try_wait_on_child(cmd, program, &mut child.0, &mut stdout) {
-                        Ok(Some(())) => {
-                            // Task done, remove the entry
-                            if is_disconnected {
-                                token.forget();
-                            }
-                            has_made_progress = true;
-                            false
-                        }
-                        Ok(None) => true, // Task still not finished, keep the entry
-                        Err(err) => {
-                            // Task fail, remove the entry.
-                            has_made_progress = true;
-                            if is_disconnected {
-                                token.forget();
+                let mut pendings_is_empty = false;
+
+                cell_update(&pendings, |mut pendings| {
+                    // Try waiting on them.
+                    retain_unordered_mut(&mut pendings, |(cmd, program, child, _token)| {
```

Comment:
It'd be nice if we could make this not O(n) per wait. I can think of ways to do that but I don't really want to land them at the same time as this. They also are plausibly overkill.

Reply:
unfortunately looping through the array is necessary, even if we use poll/epoll to poll on signalfd, unless we obtain a pidfd somehow.

Currently the method for obtaining pidfd via libstd is unstable so we would have to mess with syscalls directly to do this and I'm reluctant to try it.

Reply:
Yeah, it probably needs more thought.

Reply:
That said, looping over it only in cases where we know there are changes is a lot better than looping when there might be none.